documentation/guide/writing-persistent-objects.rst

```diff
@@ -0,0 +1,533 @@
+objects are the only ZODB-provided persistent set implementation.
+``BTree`` and ``TreeSets`` come in a number of families provided via
+different modules and differ in their internal implementations:
+
+===============  ===============  ================
+Module           Key type         Value Type
+===============  ===============  ================
+BTrees.OOBTree   object           object
+BTrees.IOBTree   integer          Object
+BTrees.OIBTree   object           integer
+BTrees.IIBTree   integer          integer
+BTrees.IFBTree   integer          float
+BTrees.LOBTree   64-bit integer   Object
+BTrees.OLBTree   object           64-bit integer
+BTrees.LLBTree   64-bit integer   64-bit integer
+BTrees.LFBTree   64-bit integer   float
+===============  ===============  ================
+
+Here's a version of the example that uses a ``TreeSet``::
+
+  import persistent
+  from BTrees.OOBTree import TreeSet
+
+  class Book(persistent.Persistent):
+
+     def __init__(self, title):
+         self.title = title
+         self.authors = TreeSet()
+
+     def add_author(self, author):
+         self.authors.add(author)
+
+.. -> src
+
+   >>> exec(src)
+   >>> db = ZODB.DB(None)
+   >>> with db.transaction() as conn:
+   ...     conn.root.book = Book("ZODB")
+   >>> conn = db.open()
+   >>> book = conn.root.book
+   >>> bool(book._p_changed)
+   False
+   >>> book.add_author('Carlos')
+   >>> bool(book._p_changed)
+   False
+   >>> bool(book.authors._p_changed)
+   True
+   >>> db.close()
+
+Properties
+==========
+
+If you implement some attributes using Python properties (or other
+types of descriptors), they are treated just like any other attributes
+by the persistence machinery.  When you set an attribute through a
+property, the object is considered changed, even if the property
+didn't actually modify the object state.
+
+Special attributes
+==================
+
+There are some attributes that are treated specially.
+
+Attributes with names starting with ``_p_`` are reserved for use by
+the persistence machinery and by ZODB.  These include (but aren't
+limited to):
+
+_p_changed
+  The ``_p_changed`` attribute has the value ``None`` if the
+  object is a :ref:`ghost <ghost-label>`, True if it's changed, an
+  False if it's not a ghost and not changed.
+
+_p_oid
+  The object's unique id in the database.
+
+_p_serial
+  The object's revision identifier also know as the object serial
+  number, also known as the object transaction id. It's a timestamp
+  and if not set has the value 0 encoded as string of 8 zero bytes.
+
+_p_jar
+  The database connection the object was accessed through.  This is
+  commonly used by database-aware application code to get hold of an
+  object's database connection.
+
+Attributes with names starting with ``_v_`` are treated as volatile.
+They aren't saved to the database.  They're useful for caching data
+that can be computed from saved data and shouldn't be saved.  They
```

Comment:
Link to/mention `zope.cachedescriptors`? Specifically `zope.cachedescriptors.property.Lazy`?